### ISSUE TITLE
Not able to install oracle_fdw for postgres 16 version on windows 10(64 bit)

### ISSUE THREAD

**Issue by chetan2211** [laurenz/oracle_fdw#609]
Hi @laurenz ,

I found the necessary files for creating the `oracle_fdw` extension in PostgreSQL 13, including `oracle_fdw.dll` from the `lib` folder and the other four files (`oracle_fdw.control`, `oracle_fdw--1.0--1.1`, `oracle_fdw--1.1--1.2`, `oracle_fdw--1.2`) from the `extension` folder. I successfully created the extension using these files.

However, for PostgreSQL 16, I can't seem to find these files in the source code, and I'm having difficulty understanding the method for creating the FDW server on Windows.

I would greatly appreciate your help with this issue.

Thank you in advance.

**Comment by laurenz** [comment]
The SQL files are there in the latest release of oracle_fdw.
There is no `oracle-fdw.dll` because I didn't build Windows binaries.
Since i have no Windows machine within reach, I cannot ready-made binaries.
You'll have to build oracle_fdw yourself; see [the documentation](https://github.com/laurenz/oracle_fdw/#building-oracle_fdw-on-windows).

**Comment by chetan2211** [comment]
Hi @laurenz  , thanks for reply. I followed the mention steps from above documentation but i am getting this log in debug folder.

  oracle_fdw.c
E:\oracle_fdw-ORACLE_FDW_2_5_0\oracle_fdw-ORACLE_FDW_2_5_0\oracle_fdw.c(8,10): fatal  error C1083: Cannot open include file: 'postgres.h': No such file or directory
  oracle_gis.c
E:\oracle_fdw-ORACLE_FDW_2_5_0\oracle_fdw-ORACLE_FDW_2_5_0\oracle_gis.c(15,10): fatal  error C1083: Cannot open include file: 'oci.h': No such file or directory
  oracle_utils.c
E:\oracle_fdw-ORACLE_FDW_2_5_0\oracle_fdw-ORACLE_FDW_2_5_0\oracle_utils.c(12,10): fatal  error C1083: Cannot open include file: 'unistd.h': No such file or directory
  oracle_msvc.c
E:\oracle_fdw-ORACLE_FDW_2_5_0\oracle_fdw-ORACLE_FDW_2_5_0\msvc\oracle_msvc.c(1,10): fatal  error C1083: Cannot open include file: 'postgres.h': No such file or directory
  Generating Code...

**Comment by laurenz** [comment]
That error means that the PostgreSQL and Oracle header files could not be found. Did you install the "devel" or "dev" package for PostgreSQL and the SDK files for Oracle?

**Comment by Piasecznik** [comment]
I have also difficulty with oracle_fdw installation on postgres 15.4
Built dll with Vistual Studio 2019 installed it, created foreign table with it but query limited to 100 rows return empty and unlimited returns:
SQL Error [58P01]: ERROR: could not load library "C:/Program Files/PostgreSQL/15/lib/oracle_fdw.dll": The specified procedure could not be found.

**Comment by Piasecznik** [comment]
I managed to come up with solution how to build working dll on Windows 2022 server using Visual Studio 2022, Oracle Instant Client 21.11 and Postgres 16.
Downloaded base oracle driver, sdk, jdbc and odbc for good measure.
Modified oracle_fdw-master\msvc\oracle_fdw.props to point to OracleClient directory and PostgreSQL directory and it was built successfully and it works.
Same dll did not want to work on 15. Had to rebuild it using same procedure.

**Comment by laurenz** [comment]
Right. You need to build it for every PostgreSQL version. `oracle_fdw.dll` built for PostgreSQL v16 will fail to run with PostgreSQL v15. This is a default property for all PostgreSQL extensions to avoid cryptic failure modes caused by ABI breaks between major versions.

**Comment by laurenz** [comment]
I think this issue is resolved; shall we close it?
